mlflow/models/evaluation/default_evaluator.py

```diff
@@ -41,20 +42,30 @@
 _DEFAULT_SAMPLE_ROWS_FOR_SHAP = 2000
 
 
-def _infer_model_type_by_labels(labels):
-    distinct_labels = set(labels)
-    for v in distinct_labels:
+def _is_categorical(values):
+    distinct_values = set(values)
+    for v in distinct_values:
         if not isinstance(v, numbers.Number):
-            return "classifier"
+            return True
         if not float(v).is_integer():
-            return "regressor"
-    if len(distinct_labels) > 1000 and len(distinct_labels) / len(labels) > 0.7:
-        return "regressor"
-    if len(distinct_labels) <= 20:
-        return "classifier"
+            return False
+    if len(distinct_values) > 1000 and len(distinct_values) / len(values) > 0.7:
+        return False
```

Comment:
How is this criteria determined? If I have a word embedding could it be classified as continuous?

Reply:
It is empirical.

> If I have a word embedding could it be classified as continuous

If a component column of a word embedding contains values all integers, it might be regarded as categorical here.

The logic is not ideal, for uncertain cases, we will treat it as categorical and compute "mode" instead of "mean".

But, I suggest, shall we let user provide information whether the feature column is categorical or continuous ?
There's no good rules to precisely infer this from feature values.

CC @BenWilson2 WDYT ?

Reply:
This is an excellent point and one that isn't universally supported in OSS ML packages but which proprietary statistical modeling packages support. 
For any 'machine data type' (e.g., int, double, string) each of these, in different scenarios and applications, could be representing data that is ordinal, nominal, or continuous. Having a hybrid typing system that represents these compound types through explicit user declaration could probably go a long way in preventing issues or perceived bugs due to complex inference logic. 
I feel like this is a broader discussion point that opens up a Pandora's box of a deep design (and incredibly out of scope for this PR), but something worth bringing up. 

The generally accepted best practice for handling data types and any inferred processing performed on them would be something akin to:

Nominal (non-ordered) [either categorical or numeric]: validate that an encoding has been performed (binary encoding, byte encoding, binning, clustering, etc.) and that is the appropriate type of encoding for the model being used based on the cardinality of a raw feature (i.e. don't do one hot encoding for a nominal value being sent to a tree-based algorithm if the count of unique values is > 30%  (placeholder; this is use-case dependent) of the data)

Ordinal (ordered) [either categorical or numeric]: generally handled by an encoder that preserves relationships through an enumeration or binning based on related characteristics of the underlying data (e.g., shirt sizes XXXS, XXS, XS -> 0 ; S, M -> 1; L -> 2; XL, XXL, XXXL -> 3....). 

I think implementing any of these validations or handling of inferred decisions based on the data without user input will be a truly daunting level of complexity. 
Can we do a 'best effort' with inference but leave it up to the users to explicitly define handling of feature types if they see that the inference isn't handling their data in the way that they expect it to?

Reply:
Agree with Ben best effort detection probably suffice here. I'd recommend to remove L50-L55. Only return True if we for sure knows it is categorical. Returns False otherwise. But we need to add Pandas type detection at the beginning L46, to make sure users can always tell us it's categorical by changing the column to "category" if that's what they want.

Reply:
Updated.
Now newest code for "category", "string", "boolean" types, `_is_categorical` return True, otherwise return False.